### DISCUSSION TITLE
🔍必要な機能に関する調査

### DISCUSSION THREAD

**Discussion by kazzyfrog** [lgtm-factory/lgtm-factory#74] [Scrap]
アプリ内で、必要な機能に関する調査結果などを、つらつらまとめていきます😎

下記などで、雑に調査したことのまとめです！
- #70 
- https://satori-image-response-practice.vercel.app/
- https://playground.lgtm-factory.pages.dev/

**Comment by kazzyfrog** [comment]
## まず、開発するサービスについて確認
- 作るのは、**GitHub上のコードレビューで使用する、LGTM画像を提供するサービス**！
- LGTM画像の雰囲気としては、LINEスタンプ・ステッカーって感じ！
- サービスの特徴としては、主に２つ
    - 雑多な写真が、無造作にアップされる感じの類似のLGTM画像サービスと違い、（一応）デザインした画像を用意するので、コンテンツの毛色が結構違う
    - 色々なデザインを追加できたり、編集できたりする、、自由な拡張性！

その他、
- HTML, CSSが書ければ、GitHub上から、誰でもデザインを追加できる、初心者〜向けのオープンソースのプロジェクトでもある！

## ユースケース
ユーザーの、メインユースケースは、大きく２つに分けられる：

- サイトを訪れて、LGTM画像を選び、URLをコピペして、そのまま使う（自分で編集せず）
- サイトを訪れて、LGTM画像を選び、自分で内容を編集して、URLコピペして、使う

**Comment by kazzyfrog** [comment]
## 事前に知っておくといいことは、２つ！

### 1.Next.js の[ImageResponse](https://nextjs.org/docs/app/api-reference/functions/image-response) 
今回の、画像生成の実装で使う機能です！🫠

> ImageResponseを使用すると、JSXとCSSを使用して動的画像を生成できます。
これは、オープングラフ画像、Twitterカードなどのソーシャルメディア画像を生成するのに便利です。

Next.jsに、標準で搭載されている、画像生成機能です！
HTML, CSS, JSX, などのコードから、画像を生成できます！👍

例えば、このコードから、下の画像を生成できます！
```jsx
<div
  style={{
    height: '100%',
    width: '100%',
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#fff',
    fontSize: 32,
    fontWeight: 600,
  }}
>
  <svg
    width="75"
    viewBox="0 0 75 65"
    fill="#000"
    style={{ margin: '0 75px' }}
  >
    <path d="M37.59.25l36.95 64H.64l36.95-64z"></path>
  </svg>
  <div style={{ marginTop: 40 }}>Hello, World</div>
</div>
```
<img width="649" alt="スクリーンショット 2024-08-18 20 25 03" src="https://github.com/user-attachments/assets/69136b1f-9492-4962-9163-fc868b10f988">

#### メモ：satoriについて
Next.js のImageResponseは、**内部で `satori` というライブラリを使用しています**！

satoriは、Next.jsと同じく、vercelが開発しているライブラリで、
HTML, CSS, JSX, などのコードを、SVGに変換するために使われます！！

また、Tailwind CSSも、問題なく変換できます！

以下、satori を使って、コードから画像の生成を試せるサイトです。
雰囲気を掴めます！
https://og-playground.vercel.app

こんな感じで、
- ベースのデザインを用意して、
- 動的な値（色とか、テキストとか、絵文字とか、、）を受け取り、
- 画像を生成して、返却する感じのLGTMサービスです！

**Reply by kazzyfrog** [reply]
### 2. Next.jsの、[Route Handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers)
今回の、画像生成APIの作成に、使う機能です！👾

Next.js は、フロントエンド開発のReactのフレームワークでありながら、
サーバーサイドのAPIも作成することができます！

つまり、Next.jsだけで、フルスタックアプリが開発できます👍

今回、生成した画像の主な使用法は、
「**画像のパス（URL）をコピーして、GitHub上に貼り付ける**」です！

なので、サーバーサイドで、画像を生成する必要があります！
厳密に言えば、
一般的なReactアプリと同じように、ブラウザ上でJSを実行し、画像を生成することも可能です。

しかし、それでは、
「**画像のパス（URL）をコピーして、GitHub上に貼り付ける**」
というのはできなくなります！

GitHub上で、JSを実行し、画像を生成することはできないためです！
なので、あらかじめ画像を生成しておいて、それをレスポンスとして返したいところです！👍

**Reply by kazzyfrog** [reply]
## まとめ

- Next.jsの、Route Handlers を使って、URLに応じた画像を生成＆返却するAPIを作成する！
- 具体的には、Next.jsの ImageResponse という機能で、画像を生成する！

**Reply by kagomen** [reply]
Route Handlersについて調べてたら、わかりやすい動画があったので共有！
https://www.youtube.com/watch?v=NZLGDva2Kbs

**Reply by ghost** [reply]
@kazzyfrog 
> URLに応じた画像を生成＆返却するAPIを作成する！

これってピンタレストみたいな感じですよね...？

**Reply by kazzyfrog** [reply]
@soraOrange 
確かに、画像が並んでいる感じなので、
サイトの雰囲気としては、ピンタレストに近いかもです！

ただ、実際の処理としては、
URLのパラメータで、色々な値を指定して画像生成する感じです！
👇
例えば、
「domain.com/lgtm-image?text=hello&bg=black」
上記のようなURLの場合は、
「hello」という単語が入った、背景が黒の画像が生成される！

みたいな感じです！👍

**Reply by ghost** [reply]
@kazzyfrog なるほど！理解しました！
ありがとうございます！

**Comment by kazzyfrog** [comment]
## 「LGTM画像サービス」
ちなみに、検索で出てくる「LGTM画像サービス」は、ざっとこんな感じです！

このプロジェクトに関する、
アイデアを与えてくれた、サービスたちです！

時間のある時に、どんな感じなのか、触ってみてくださいー👍

[LGTMoon](https://lgtmoon.herokuapp.com)
一般的な、「LGTM画像サービス」

[LooksToMe](https://looks-to.me)
上記と同じ感じ！

[LGTM GEN](https://lgtmgen.com/)
上記サイトなどよりは、ログイン登録必須にして、変な画像アップされないようにした、らしい
> ある程度治安のいいLGTM画像を生成できるサービスを作りました。

[LGTMeow](https://lgtmeow.com)
猫の画像限定！

**Reply by kazzyfrog** [reply]
上記を見るとわかる通り、一般的なLGTM画像サービスは、
- 自由に画像をアップロードしたり、画像検索から引っ張ってくる
- その画像に、「LGTM」の文字を載せる
- URLをコピーして、すぐに使える！

こんな感じです！🧐

**Comment by kazzyfrog** [comment]
## これらを踏まえて、ここ数日で調査したことは、、

- まず、ImageResponseや、内部で利用されているsatori というライブラリを、試しに触って、簡単な画像を生成してみた！
    - 基本は、JSX（HTML, CSS）のコードから画像を生成するので、簡単なWebデザインって感じで、自由度は高い
    - フォントや、生成する画像のサイズなども、オプションとして自由にカスタマイズできる

実際の、ImageResponse のコードはこんな感じです！
```jsx
import { ImageResponse } from 'next/og'
 
new ImageResponse(
  // 第１引数に、画像に変換したいJSXを受け取る
  element: ReactElement,
  
  // 第２引数に、サイズやフォントなどのオプションを受け取る
  options: {
    width?: number = 1200
    height?: number = 630
    emoji?: 'twemoji' | 'blobmoji' | 'noto' | 'openmoji' = 'twemoji',
    fonts?: {
   ~~
  }
)
```

これで画像が生成される！

**Reply by kagomen** [reply]
コメントつきでソースコード載せてるので、良ければ参考にしてください！
https://github.com/kagomen/satori-image-response-practice
